app/hooks/usePermissions.ts

```diff
@@ -0,0 +1,84 @@
+import { useState, useEffect } from 'react';
+import { useSelector } from 'react-redux';
+import { dequal } from 'dequal';
+import { Subscription } from 'rxjs';
+
+import { TSupportedPermissions } from '../reducers/permissions';
+import { IApplicationState } from '../definitions';
+import { getUserSelector } from '../selectors/login';
+import database from '../lib/database';
+import log from '../utils/log';
+
+type TPermissionState = (boolean | undefined)[];
+
+export function usePermissions(permissions: TSupportedPermissions[], rid?: string) {
+	const [permissionsState, setPermissionsState] = useState<TPermissionState>([]);
+	const [roomRoles, setRoomRoles] = useState<string[]>([]);
+	let subscription: Subscription | null = null;
```

Comment:
If you declare a simple javascript variable yourself, it will be updated on each render. Refs are used when you need to persist some value during re-renders

```suggestion
	const subscription = useRef<Subscription | null>();
```